packages/js/src/cache/notifications-cache.ts

```diff
@@ -0,0 +1,254 @@
+import { InMemoryCache } from './in-memory-cache';
+import type { Cache } from './types';
+import type { ListNotificationsArgs, ListNotificationsResponse, Notification } from '../notifications';
+import { NotificationEvents, NotificationsEvents, NovuEventEmitter } from '../event-emitter';
+import type { NotificationFilter } from '../types';
+import { areTagsEqual, isSameFilter } from '../utils/notification-utils';
+
+const getCacheKey = (args: ListNotificationsArgs): string => {
+  return JSON.stringify(args);
+};
+
+const getFilterKey = ({ tags, read, archived }: Pick<ListNotificationsArgs, 'tags' | 'read' | 'archived'>): string => {
+  return JSON.stringify({ tags, read, archived });
+};
+
+const getFilter = (key: string): NotificationFilter => {
+  return JSON.parse(key);
+};
+
+// these events should update the notification in the cache
+const notificationEvents: NotificationEvents[] = [
+  'notification.read.pending',
+  'notification.read.resolved',
+  'notification.unread.pending',
+  'notification.unread.resolved',
+  'notification.complete_action.pending',
+  'notification.complete_action.resolved',
+  'notification.revert_action.pending',
+  'notification.revert_action.resolved',
+];
+
+// these events should remove the notification from the cache
+const notificationEventsThatRemove: NotificationEvents[] = [
+  'notification.archive.pending',
+  'notification.unarchive.pending',
+];
+
+// these events should update the notifications in the cache
+const notificationsEvents: NotificationsEvents[] = [
+  'notifications.read_all.pending',
+  'notifications.read_all.resolved',
+];
+
+// these events should remove the notifications from the cache
+const notificationsEventsThatRemove: NotificationsEvents[] = [
+  'notifications.archive_all.pending',
+  'notifications.archive_all_read.pending',
+];
```

Comment:
❓ Does this really make sense here?

I don't think this is the place to do this. If a notification is marked as read and a new call to the fetch function is made, it should not include the notification that was marked as read. Think of it like switching to another tab and going back again. A call is made again and the notification should disappear, although no network requests were made. At least that's what makes sense to me, dunno how doable it is. Let's discuss.

Reply:
❓ Does this really make sense here?

I don't think this is the place to do this. If a notification is marked as read and a new call to the fetch function is made, it should not include the notification that was marked as read. Think of it like switching to another tab and going back again. A call is made again and the notification should disappear, although no network requests were made. At least that's what makes sense to me, dunno how doable it is. Let's discuss.

Reply:
@desiprisg 

**Behavior:**
It all depends on the top filter that would be applied in the header we have three: `unread + read`, `unread`, `archived`. From what I talked with Nikolay and @SokratisVidros we agreed and wanted to copy the same behaviour from Notion, meaning:
- read/unread actions keep the notification in the list
  - with the `unread + read` filter you will always see it even after refresh
  - with the `unread` filter it will disappear after refresh
- archive/unarchive actions remove notification from the list
  - no matter what filter is applied 

**Cache:**
We talked about this case in the cache meeting that we had: the `Notification` class provides the functionality to mark it as read/unread/archive/unarchive. For example `notification.read()` will optimistically update the notification fields and the value in the cache and will call API with a response updating the cache again (because API is the source of truth). This is what we have agreed, logic in JS SDK.

We can't do a refetch after it's marked as read, because there might be new notifications that will shift the whole list so we would need to refetch all pages.

Reply:
Let’s stick to Notion-like behavior for now. 

The cache powers the UI. Changes to individual notifications are applied to the list but notifications are not removed from the list unless the cache is invalidated.

Gmail does the same thing, you filter by unread then you mark an email as read, it’s not bold anymore but it is in the unread list until the next reload.

Reply:
Ok so in essence, the update happens to the notification only, not any lists. Fine by me if that's how we envisioned it. I can see the benefits in both!

Looks good then!